### ISSUE TITLE
Should Arrow keys cycle through options?

### ISSUE THREAD

**Issue by terracoda** [phetsims/molecules-and-light#153]
From #143, @terracoda noticed that

> Just tested in Safari 10.0.3, and I see that the Arrow keys do not cycle through the list of molecules, or the list of lights like they do in Firefox. I think cycling through is the expected behaviour, but I will have to check.

**Comment by jessegreenberg** [comment]
@terracoda can you please check what Safari does in this example?

https://www.w3schools.com/html/tryit.asp?filename=tryhtml_form_radio

**Comment by terracoda** [comment]
@jessegreenberg, Safari does not cycle through on the w3schools example, but in this example, where `role="radiogroup"` is being used, Safari cycles through.

According to the ARIA Practices radiobutton items in a radiogroup should cycle through.
https://www.w3.org/TR/wai-aria-practices-1.1/#radiobutton

Menu items in a menuliist would also cycle through. I am assuming they will be radiobutton items. Do you agree that they are each a group of radiobuttons?

**Comment by jessegreenberg** [comment]
Thanks @terracoda it is the difference between

```html
<fieldset>
  <input type="radio">
</fieldset>
```

and
```html 
<div role="radiogroup">
  <div role="radiobutton"></div>
</div>
```

We went with the first option because ARIA didn't seem necessary. The bottom one will cycle through because there is a custom script to do so. The second one cycles in Safari because there is custom JavaScript for the behavior.

**Comment by jessegreenberg** [comment]
@terracoda which markup would you rather have, and how important is it that radios cycle in Safari? We could switch to the ARIA approach, but I would guess it might take 2-6 hours to do a custom solution with ARIA that would involve

- Modifying sun/RadioButtonGroup to handle arrow keys on our own.
- Testing role="radiogroup" and role="radiobutton" to see if they are supported in NVDA and JAWS.

Since the markup is sound with ```<input type="radio>``` and Safari chooses not to cycle normal radio buttons I would recommend sticking with typical HTML.

**Comment by terracoda** [comment]
Hmm, does `fieldset` have a mandatory default role? Maybe we can just add the roles without changing the html?

**Comment by terracoda** [comment]
Ideally, we want the sim to work the same in all browsers if possible?

**Comment by jessegreenberg** [comment]
@terracoda we can also add some ARIA to the current markup, but it seems more likely that the cycling behavior is coming from the associated JavaScript:
https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/examples/radio/js/radio.js

**Comment by terracoda** [comment]
Well, fixing disparities between user agents is one of the arguments for/ goals of Web Standards. 

This is likely a Safari bug. It is pretty normal to cycle through groups of things (menu items, radio button items, listbox items).

**Comment by terracoda** [comment]
Re https://github.com/phetsims/molecules-and-light/issues/153#issuecomment-326106131, yes, I think you are right. The cycling behaviour is built into the JavaScript because user agents do not all do the same thing, i.e., implement the specification as described.

Cycling in Safari likely won't come automatically by just adding the aria roles.

**Comment by terracoda** [comment]
@jessegreenberg, I think it is worth investigating how well supported, the role "radiogroup" is in browsers and screen readers. The HTML `fieldset` element does not actually sound consistent across screen readers and all items within the `fieldset` are counted as items, rather than just the actual interactive items.

It's possible that 
```html 
<div role="radiogroup">
</div>
``` 
provides a better experience all around.

It is also possible that using an un-ordered list to group the items may work more consistently than either `fieldset` or `radiogroup`. However, whatever code we use, the interaction question remains, should the Arrow keys automatically cycle through the items in a group of items of the same type. I think this would be the preferred behaviour recommended by the ARIA Practices documentation, though it may vary depending on the type of the items in the group.

I

**Comment by terracoda** [comment]
@jessegreenberg, before making a big change in sun/RadioButtonGroup, maybe I could just provide some simple HTML snippets.

**Comment by terracoda** [comment]
Here's another example using a `ul` and `aria-activedescendent`:
https://www.w3.org/TR/wai-aria-practices/examples/radio/radio-2/radio-2.html

**Comment by jessegreenberg** [comment]
> Well, fixing disparities between user agents is one of the arguments for/ goals of Web Standards.
This is likely a Safari bug. It is pretty normal to cycle through groups of things (menu items, radio button items, listbox items).

Safari is going against the grain, but I question whether it is PhET's responsibility to ensure consistency. To me this seems like a case where we adhere to standards and assume the browser takes care of the rest. 

> I think it is worth investigating how well supported, the role "radiogroup" is in browsers and screen readers.

I agree, definitely worth looking into.  We initially avoided it because the general recommendation is to "avoid ARIA where possible" because our experience has been that HTML performs better than ARIA.

**Comment by terracoda** [comment]
@jessegreenberg 
> I agree, definitely worth looking into. We initially avoided it because the general recommendation is to "avoid ARIA where possible" because our experience has been that HTML performs better than ARIA.

Hmm, I feel it is a bit more nuanced than that. Yes, we should use HTML wherever possible first, and only use ARI when HTML can't do the job on its own. The ARIA specification came into being because HTML did not have the built-in semantics to describe all states, roles, and properties needed to make web applications more accessible. Then, HTML (HTML5) started adding more native elements to cover more semantics. We are now in a landscape of change, where new semantic HTML elements are not always supported by browsers, and where sometimes ARIA attributes work better. The reverse is also true. And sometimes things work best if both are used together because some AT supports ARIA better while a browser may support HTML better.

There is no `radiogroup` element. It only exists as an ARIA role specifically to group elements with `role=radio` or `input` elements of `type="radio"` which should get the the `role="radio"` by default.

I think I have seen it before, but I just read the 5 rules of ARIA here,[ Using Aria](https://www.w3.org/TR/using-aria/). It is a nice succinct guidance. It is in "draft" form.

**Comment by terracoda** [comment]
More specifically, how should we go about testing some code samples that can give us the most accessible radio button lists in order to create an inclusive design pattern that serves PhET sims. There are several options:

1. ` fieldset `element as a grouping option -  Gets a group role by default. Drawback AT counts all elements inside the fieldset as items. May be readout inconsistently by AT.
2. `ul` element as a grouping option - this might be the simplest option. The ARIA`role=radiogroup` is allowed on a `ul` if we want to use it.
3. `div` element with `role=radiogroup` as a grouping option. A role would be required if we use a div.

As for the items inside, I think it is likely best to go with the `input type="radio"`, the native HTML option.

I can provide the HTML, but I would need someone else to provide the JavaScript when needed.

**Comment by terracoda** [comment]
In the [Using ARIA document section 2.14](https://www.w3.org/TR/using-aria/#html-aria-gaps), the is a list of ARIA roles not native to HTML.  The roles `radiogroup` and `radio` are not on the list, so in theory we should be able to communicate these roles natively in HTML.

**Comment by terracoda** [comment]
Maybe this issue should be moved to the a11y-research repo?

**Comment by jessegreenberg** [comment]
> Maybe this issue should be moved to the a11y-research repo?

Good idea @terracoda, this seems more general than specific to molecules-and-light.

**Comment by jessegreenberg** [comment]
In the last a11y meeting we decided that we will allow Safari to use its default behavior for radio  buttons and not cycle through from  bottom to top.
